### PR TITLE
Extract the ES usage analyzer from the ES minifier crate

### PR DESCRIPTION
**Pull request by alexkirsz**
**Description:**

This PR extracts the EcmaScript usage analyzer from the ES minifier crate into its own crate, to make it reusable by third parties. Notably, we plan on using this in Turbopack to implement tree shaking.

**BREAKING CHANGE:**

The `swc_ecma_minifier::marks::Marks` type is now exported from `swc_ecma_usage_analyzer::marks::Marks` instead. It might make sense to also re-export it from the `swc_ecma_minifier` crate.

**Related issue (if exists):**

Fixes WEB-269

### CODE REVIEW

**Review by wbinnssmith — DISMISSED** [review]

File: crates/swc_ecma_usage_analyzer/src/analyzer/mod.rs [hunk 1]
```diff
@@ -234,7 +234,7 @@ enum RecursiveUsage {
 
 /// Analyzed info of a whole program we are working on.
 #[derive(Debug, Default)]
-pub(crate) struct ProgramData {
+pub struct ProgramData {
```
Comment (kdy1) [inline crates/swc_ecma_usage_analyzer/src/analyzer/mod.rs:237]: Are you going to use this type directly?

Comment (kdy1) [reply crates/swc_ecma_usage_analyzer/src/analyzer/mod.rs:237]: I prefer storing this type in `swc_ecma_minifier`, to reduce breaking changes.

**Review by kdy1 — COMMENTED** [review]
I'm bit worried about the number of breaking changes

File: crates/swc_ecma_usage_analyzer/src/analyzer/mod.rs (hunk 1 above)
Comment (alexkirsz) [reply crates/swc_ecma_usage_analyzer/src/analyzer/mod.rs:237]: The current definition of `UsageAnalyzer` depends on it: `UsageAnalyzer<S = ProgramData>`. But I agree this makes more sense to live in `swc_ecma_minifier` for now so I'll move it and remove the default type.

From what I can tell the only breaking change is the `marks::Marks` type changing crates, and we can always re-export it from `swc_ecma_minifier`.

Comment (kdy1) [reply crates/swc_ecma_usage_analyzer/src/analyzer/mod.rs:237]: Yeah, I think it's fine. Thank you!

File: crates/swc_ecma_usage_analyzer/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+pub mod alias;
+pub mod analyzer;
+pub mod marks;
+pub mod util;
+
+#[cfg(test)]
```
Comment (kdy1) [inline crates/swc_ecma_usage_analyzer/src/lib.rs:6]: Can you remove this?

File: crates/swc_ecma_usage_analyzer/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+mod tests {
```
Comment (kdy1) [inline crates/swc_ecma_usage_analyzer/src/lib.rs:7]: Can you remove this module?

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_minifier

**Pull request merged**